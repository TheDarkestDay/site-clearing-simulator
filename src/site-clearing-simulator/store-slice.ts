import { createSlice } from "@reduxjs/toolkit";

export type SiteCellType = "o" | "t" | "r" | "T" | "B" | "C";

export enum BulldozerDirection {
  Up = 1,

  Right = 2,

  Down = 3,

  Left = 4,
}

export type SiteClearingSimulatorState = {
  map: SiteCellType[][];
  isStarted: boolean;
  isStopped: boolean;
  isPreservedTreeRemoved: boolean;
  bulldozerPosition: {
    x: number;
    y: number;
  };
  bulldozerDirection: BulldozerDirection;
  fuelUsed: number;
};

// ["o", "o", "t", "o", "o", "o", "o", "o", "o", "o"],
//     ["o", "o", "o", "o", "o", "o", "o", "T", "o", "o"],
//     ["r", "r", "r", "o", "o", "o", "t", "T", "o", "o"],
//     ["r", "r", "r", "r", "o", "o", "o", "o", "o", "o"],
//     ["r", "r", "r", "r", "r", "t", "o", "o", "o", "o"]

const initialState: SiteClearingSimulatorState = {
  map: [],
  isStarted: false,
  isStopped: false,
  bulldozerDirection: BulldozerDirection.Right,
  bulldozerPosition: {
    x: -1,
    y: -1,
  },
  isPreservedTreeRemoved: false,
  fuelUsed: 0,
};

const siteClearingSimulatorSlice = createSlice({
  name: "siteClearingSimulator",
  initialState,
  reducers: {
    startSimulation(state, action) {},
    stopSimulation() {},
    moveForward(state) {
      return state;
    },
    rotateLeft(state) {
      return state;
    },
    rotateRight(state) {
      return state;
    },
    setMap(state, action) {
      return state;
    },
  },
});

export const {
  startSimulation,
  stopSimulation,
  moveForward,
  rotateLeft,
  rotateRight,
  setMap,
} = siteClearingSimulatorSlice.actions;

export default siteClearingSimulatorSlice.reducer;
