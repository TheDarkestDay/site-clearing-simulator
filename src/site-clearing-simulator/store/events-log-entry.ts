import { LogEntryType } from './log-entry-type';

export type EventsLogEntry = {
  type: LogEntryType;
  description: string;
};