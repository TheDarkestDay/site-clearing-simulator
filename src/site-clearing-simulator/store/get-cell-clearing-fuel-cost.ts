import { SiteCellType } from './store-slice';

export const getCellClearingFuelCost = (cellType: SiteCellType): number => {
  switch (cellType) {
    case 'o':
    case 'C':
      return 1;
    case 'r':
    case 't':
      return 2;
    default:
      throw new Error('Clearing of a cell with bulldozer or preservable tree is not allowed');
  }
};