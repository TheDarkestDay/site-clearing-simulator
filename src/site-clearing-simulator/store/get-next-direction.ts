import { BulldozerDirection } from './store-slice';

export const getNextDirection = (direction: BulldozerDirection, turnDirection: 'left' | 'right'): BulldozerDirection => {
  const directions = [
    BulldozerDirection.Up,
    BulldozerDirection.Right,
    BulldozerDirection.Down,
    BulldozerDirection.Left,
  ];

  const directionDelta = turnDirection === 'right' ? 1 : -1;
  const newDirectionIndex = directions.indexOf(direction) + directionDelta;

  if (newDirectionIndex < 0) {
    return BulldozerDirection.Left;
  }

  if (newDirectionIndex >= directions.length) {
    return BulldozerDirection.Up;
  }

  return directions[newDirectionIndex];
}