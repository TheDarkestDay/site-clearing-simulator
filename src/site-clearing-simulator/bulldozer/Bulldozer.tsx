import { useSelector } from 'react-redux';
import { getBulldozerDirection } from '../store/selectors';
import { BulldozerDirection } from '../store/store-slice';
import styles from './styles.module.css';

const getBulldozerDirectionClass = (direction: BulldozerDirection) => {
  switch (direction) {
    case BulldozerDirection.Down:
      return styles.bulldozerHeadingDown;
    case BulldozerDirection.Right:
      return styles.bulldozerHeadingRight;
    case BulldozerDirection.Up:
      return styles.bulldozerHeadingUp;
  }
};

export const Bulldozer = () => {
  const bulldozerDirection = useSelector(getBulldozerDirection);
  const bulldozerClassName = `${styles.bulldozer} ${getBulldozerDirectionClass(bulldozerDirection)}`

  return (
    <div className={bulldozerClassName}>
      <span>ðŸšœ</span>
    </div>
  );
};