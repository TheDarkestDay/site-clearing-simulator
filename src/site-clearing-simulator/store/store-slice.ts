import { createSlice } from "@reduxjs/toolkit";
import { dir } from 'console';
import { EventsLogEntry } from './events-log-entry';
import { LogEntryType } from './log-entry-type';

export type SiteCellType = "o" | "t" | "r" | "T" | "B" | "C";

export type Point2D = {
  x: number;
  y: number;
}

export enum BulldozerDirection {
  Up = 'UP',

  Right = 'RIGHT',

  Down = 'DOWN',

  Left = 'LEFT',
}

export type SiteClearingSimulatorState = {
  map: SiteCellType[][];
  isStarted: boolean;
  isStopped: boolean;
  isPreservedTreeRemoved: boolean;
  bulldozerPosition: Point2D;
  bulldozerDirection: BulldozerDirection;
  fuelUsed: number;
  eventsLog: EventsLogEntry[];
};

export const initialState: SiteClearingSimulatorState = {
  map: [],
  isStarted: false,
  isStopped: false,
  bulldozerDirection: BulldozerDirection.Right,
  bulldozerPosition: {
    x: 0,
    y: -1,
  },
  isPreservedTreeRemoved: false,
  fuelUsed: 0,
  eventsLog: [],
};

const getNextPoint = ({ x, y }: Point2D, direction: BulldozerDirection): Point2D => {
  switch (direction) {
    case BulldozerDirection.Up:
      return {
        x: x - 1,
        y,
      }
    case BulldozerDirection.Right:
      return {
        x,
        y: y + 1,
      }
    case BulldozerDirection.Down:
      return {
        x: x + 1,
        y
      }
    case BulldozerDirection.Left:
      return {
        x,
        y: y - 1
      }
  }
};

const getNextDirection = (direction: BulldozerDirection, turnDirection: 'left' | 'right'): BulldozerDirection => {
  const directions = [
    BulldozerDirection.Up,
    BulldozerDirection.Right,
    BulldozerDirection.Down,
    BulldozerDirection.Left,
  ];

  const directionDelta = turnDirection === 'right' ? 1 : -1;
  const newDirectionIndex = directions.indexOf(direction) + directionDelta;

  if (newDirectionIndex < 0) {
    return BulldozerDirection.Left;
  }

  if (newDirectionIndex >= directions.length) {
    return BulldozerDirection.Up;
  }

  return directions[newDirectionIndex];
}

const getCellClearingFuelCost = (cellType: SiteCellType): number => {
  switch (cellType) {
    case 'o':
    case 'C':
      return 1;
    case 'r':
    case 't':
      return 2;
    default:
      throw new Error('Clearing of a cell with bulldozer or preservable tree is not allowed');
  }
};

const isPointOutOfBounds = ({x, y}: Point2D, map: SiteCellType[][]): boolean => {
  const [firstRow] = map;
  const maxX = map.length;
  const maxY = firstRow.length;

  return x < 0 || x >= maxX || y < 0 || y >= maxY; 
};

const siteClearingSimulatorSlice = createSlice({
  name: "siteClearingSimulator",
  initialState,
  reducers: {
    startSimulation(_, action) {
      return {
        ...initialState,
        map: action.payload,
        isStarted: true,
      }
    },
    resetSimulation() {
      return initialState;
    },
    stopSimulation(state) {
      state.isStopped = true;

      state.eventsLog.push({
        type: LogEntryType.Stop,
        description: 'Simulation was stopped by the user'
      });
    },
    moveForward(state) {
      const {bulldozerPosition, bulldozerDirection} = state;
      const nextPoint = getNextPoint(bulldozerPosition, bulldozerDirection);

      if (isPointOutOfBounds(nextPoint, state.map)) {
        state.isStopped = true;
        state.eventsLog.push({
          type: LogEntryType.Stop,
          description: 'Simulation stopped: bulldozer left the site boundaries'
        });

        return state;
      }

      const {x: nextX, y: nextY} = nextPoint;
      const cellToClear = state.map[nextX][nextY];
      if (cellToClear === 'T') {
        state.isPreservedTreeRemoved = true;
        state.isStopped = true;

        state.eventsLog.push({
          type: LogEntryType.Stop,
          description: 'Simulation stopped: an attempt to remove preservable tree was performed'
        });

        return state;
      }

      state.fuelUsed += getCellClearingFuelCost(cellToClear);

      const {x: previousX, y: previousY} = bulldozerPosition;
      state.map[previousX][previousY] = 'C';

      state.map[nextX][nextY] = 'B';
      state.bulldozerPosition.x = nextX;
      state.bulldozerPosition.y = nextY; 

      state.eventsLog.push({
        type: LogEntryType.MoveTo,
        description: `Move to ${nextX}-${nextY}`
      });

      const isThereAnyLandToClear = state.map
        .reduce((total, row) => total.concat(row), [])
        .some((cell) => cell === 'o' || cell === 'r' || cell === 't');

      if (!isThereAnyLandToClear) {
        state.isStopped = true;

        state.eventsLog.push({
          type: LogEntryType.Stop,
          description: 'Simulation stopped: there are no more fields to clear'
        });
      }
    },
    rotateLeft(state) {
      state.bulldozerDirection = getNextDirection(state.bulldozerDirection, 'left');

      state.eventsLog.push({
        type: LogEntryType.RotateLeft,
        description: 'Rotate left',
      });
    },
    rotateRight(state) {
      state.bulldozerDirection = getNextDirection(state.bulldozerDirection, 'right');;

      state.eventsLog.push({
        type: LogEntryType.RotateRight,
        description: 'Rotate right',
      });
    },
  },
});

export const {
  startSimulation,
  resetSimulation,
  stopSimulation,
  moveForward,
  rotateLeft,
  rotateRight,
} = siteClearingSimulatorSlice.actions;

export default siteClearingSimulatorSlice.reducer;
