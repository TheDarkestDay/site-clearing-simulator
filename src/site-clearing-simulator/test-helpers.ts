import { SiteClearingSimulatorState, BulldozerDirection } from './store/store-slice';

export const getTestState = (
  overrides: Partial<SiteClearingSimulatorState> = {}
): SiteClearingSimulatorState => {
  return {
    map: [],
    isStarted: false,
    isStopped: false,
    bulldozerDirection: BulldozerDirection.Right,
    bulldozerPosition: {
      x: 0,
      y: -1,
    },
    isPreservedTreeRemoved: false,
    fuelUsed: 0,
    eventsLog: [],
    ...overrides,
  };
};