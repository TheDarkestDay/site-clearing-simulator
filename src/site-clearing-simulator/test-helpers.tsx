import { createStore } from '@reduxjs/toolkit';
import { create } from 'react-test-renderer';
import React from 'react';
import { Provider } from 'react-redux';
import { GlobalState, rootReducer } from '../store';
import { SiteClearingSimulatorState, BulldozerDirection, initialState } from './store/store-slice';

export const getTestState = (
  overrides: Partial<SiteClearingSimulatorState> = {}
): SiteClearingSimulatorState => {
  return {
    map: [],
    isStarted: false,
    isStopped: false,
    bulldozerDirection: BulldozerDirection.Right,
    bulldozerPosition: {
      x: 0,
      y: -1,
    },
    isPreservedTreeRemoved: false,
    fuelUsed: 0,
    eventsLog: [],
    ...overrides,
  };
};

export const renderWithStore = (component: React.ReactNode, state: Partial<GlobalState['siteClearingSimulator']> = {}) => {
  const testStore = createStore(rootReducer, {
    siteClearingSimulator: {
      ...initialState,
      ...state,
    }
  });
  const consumer = (
    <Provider store={testStore}>
      {component}
    </Provider>
  );

  return create(consumer);
}