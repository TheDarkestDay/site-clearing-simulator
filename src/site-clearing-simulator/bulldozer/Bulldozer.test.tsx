import { BulldozerDirection } from '../store/store-slice';
import { renderWithStore } from '../test-helpers';
import { Bulldozer } from './Bulldozer';

describe('Bulldozer', () => {
  it('should match snapshot for bulldozer direction up', () => {
    const wrapper = renderWithStore(<Bulldozer />, {
      bulldozerDirection: BulldozerDirection.Up,
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot for bulldozer direction right', () => {
    const wrapper = renderWithStore(<Bulldozer />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot for bulldozer direction down', () => {
    const wrapper = renderWithStore(<Bulldozer />, {
      bulldozerDirection: BulldozerDirection.Down,
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot for bulldozer direction left', () => {
    const wrapper = renderWithStore(<Bulldozer />, {
      bulldozerDirection: BulldozerDirection.Left,
    });

    expect(wrapper).toMatchSnapshot();
  });
});