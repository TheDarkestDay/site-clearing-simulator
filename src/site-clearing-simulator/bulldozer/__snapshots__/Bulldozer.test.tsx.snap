// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bulldozer should match snapshot for bulldozer direction down 1`] = `
<div
  className="bulldozer bulldozerHeadingDown"
>
  <span>
    ðŸšœ
  </span>
</div>
`;

exports[`Bulldozer should match snapshot for bulldozer direction left 1`] = `
<div
  className="bulldozer "
>
  <span>
    ðŸšœ
  </span>
</div>
`;

exports[`Bulldozer should match snapshot for bulldozer direction right 1`] = `
<div
  className="bulldozer bulldozerHeadingRight"
>
  <span>
    ðŸšœ
  </span>
</div>
`;

exports[`Bulldozer should match snapshot for bulldozer direction up 1`] = `
<div
  className="bulldozer bulldozerHeadingUp"
>
  <span>
    ðŸšœ
  </span>
</div>
`;
